Undo snackbar

andOrlando opened the pull request:
Closes #171
Adds dynamic snackbar support and an undo snackbar that pops up whenever you delete an assignment
snackbars being basically notification messages

![image](https://user-images.githubusercontent.com/59105749/102719877-97a00c00-42be-11eb-8f7c-c5909da763b4.png)

andOrlando committed:
adds snackbar support

adds the snackbar class

andOrlando committed:
adds undo snackbar and fixes bugs

- adds colors for undo snackbar (to be edited)
- improves resize logic for snackbar at low screensizes
- When assignment is deleted, replaces item with a placeholder that's ignored by tabulator which gets deleted if the snackbar times out
- adds a bunch of semicolons
- better document snackbar options
- adds timeout and funciton on timeout
- improves instant .show() logic
- improves oncick things
- fixes lots of other bugs

andOrlando committed:
improves colors

andOrlando committed:
fixes placeholder deletion

- makes placeholder deletion actually delete from the list
- also adds some semicolons

andOrlando committed:
code improvements

adds some semicolons, deletes test function

psvenk commented on public/js/home.js:
```diff
@@ -578,7 +578,19 @@ let assignmentsTable = new Tabulator("#assignmentsTable", {
+                    buttonClick: () => {replaceAssignmentFromID({assignment_id: data["assignment_id"], placeholder: true}, data, selected_class_i);},
+                    timeout: 7500,
+                    timeoutFunction: () => {removeAssignmentFromID(data["assignment_id"], selected_class_i);},
+                    bodyClick: () => {removeAssignmentFromID(data["assignment_id"], selected_class_i);}
```
These arrow functions can be written more simply:
```suggestion
                    buttonClick: () => replaceAssignmentFromID({ assignment_id: data["assignment_id"], placeholder: true }, data, selected_class_i),
                    timeout: 7500,
                    timeoutFunction: () => removeAssignmentFromID(data["assignment_id"], selected_class_i),
                    bodyClick: () => removeAssignmentFromID(data["assignment_id"], selected_class_i),
```

psvenk commented on public/js/home.js:
```diff
@@ -1181,6 +1193,223 @@ function closeSideNav() {
+     * returns the snackbar object
+     */
+    make() {
+
```
Remove unnecessary blank line at beginning of function body
```suggestion
```

psvenk commented on public/js/home.js:
```diff
@@ -1181,6 +1193,223 @@ function closeSideNav() {
+     * if it's not already hidden, hides it unless override is true
+     */
+    destroy() {
+
```
Remove unnecessary blank line at beginning of function body
```suggestion
```

psvenk commented on public/js/home.js:
```diff
@@ -1181,6 +1193,223 @@ function closeSideNav() {
+     * also creates its reference in snackbars
+     */
+    createID() {
+        
```
Remove unnecessary blank line at beginning of function body
```suggestion
```

psvenk commented on public/js/home.js:
```diff
@@ -1181,6 +1193,223 @@ function closeSideNav() {
 
+
+class Snackbar {
+
```
Remove unnecessary blank line at beginning of class body
```suggestion
```

psvenk commented on public/js/home.js:
```diff
@@ -578,7 +578,19 @@ let assignmentsTable = new Tabulator("#assignmentsTable", {
             align: "center",
             cellClick: function(e, cell) {
-                cell.getRow().delete();
+
```
Remove unnecessary blank line at beginning of function body
```suggestion
```

psvenk commented on public/js/buttonFunctions.js:
```diff
@@ -128,7 +151,9 @@ let updateGradePage = function() {
-  assignmentsTable.replaceData(currentTableData.currentTermData.classes[selected_class_i].assignments);
+  assignmentsTable.replaceData(currentTableData.currentTermData.classes[selected_class_i].assignments.filter((obj) => {
+    return obj.placeholder !== true
+  }));
```
This arrow function can be simplified using [implicit returns][1] (when braces are omitted) and [object destructuring syntax][2] (and brought within 80 characters per line):
```suggestion
  assignmentsTable.replaceData(currentTableData.currentTermData
	  .classes[selected_class_i].assignments
	  .filter(({ placeholder }) => !placeholder)
  );
```

[1]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions#Basic_syntax
[2]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment

psvenk commented on public/js/buttonFunctions.js:
```diff
@@ -19,6 +19,29 @@ let newAssignment = function() {
+      index = i;
+      break;
+    }
+  }
```
I think the reason why using regular `indexOf` does not work here is because it expects perfect equality between objects and not just structural equality or equality of assignment IDs. In any case, this can be refactored in a way that avoids a C-style `for` loop and is more readable (using [ES6 features](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/findIndex)):
```suggestion
  const index = assignments.findIndex(({ assignment_id }) =>
    assignment_id === oldData.assignment_id
  );
```

psvenk commented on public/js/buttonFunctions.js:
```diff
@@ -19,6 +19,29 @@ let newAssignment = function() {
+  const assignments = currentTableData.currentTermData.classes[classID].assignments;
+  const newArray = assignments.filter((obj) => {
+    return obj["assignment_id"] !== id;
+  });
```
This arrow function can be simplified:
```suggestion
  const newArray = assignments.filter(({ assignment_id }) =>
    assignment_id !== id
  );
```

psvenk commented on public/js/buttonFunctions.js:
```diff
@@ -19,6 +19,29 @@ let newAssignment = function() {
+    return obj["assignment_id"] !== id;
+  });
+  assignments.length = 0;
+  assignments.push.apply(assignments, newArray);
```
Instead of using ```Function.prototype.apply```, this can be rewritten in ES6 as:
```suggestion
  assignments.length = 0;
  assignments.push(...newArray);
```

notrodes commented:
I think this is a great canadate for using webcomponants to compartementalize the compleity of this addition

psvenk commented:
I wonder if we could make it more intuitive to undo deleting an assignment by not actually deleting the assignment's row from the table — we could grey out the row, exclude it from calculations, and replace the "x" button with a button to restore the assignment. The snackbar code could still be useful otherwise. @andOrlando Your thoughts?

andOrlando committed:
Apply suggestions from code review

Co-authored-by: psvenk <45520974+psvenk@users.noreply.github.com>

psvenk commented on public/js/home.js:
```diff
@@ -1181,6 +1192,220 @@ function closeSideNav() {
+
+        //sets the body onclick listener which just destroys it by default
+        const bodyOnClickFunction = this.bodyClick !== undefined ? () => {this.bodyClick();} : () => {};
+        const destroyFromBody = this.destroyWhenBodyClicked ? () => {this.destroy();} : () => {};
```
Simplify arrow functions
```suggestion
        const bodyOnClickFunction = this.bodyClick !== undefined ? () => this.bodyClick() : () => {};
        const destroyFromBody = this.destroyWhenBodyClicked ? () => this.destroy() : () => {};
```

psvenk commented on public/js/home.js:
```diff
@@ -1181,6 +1192,220 @@ function closeSideNav() {
+
+            //sets the button onclick listener which runs the given funtion and destroys the snackar by default
+            const destroyFromButton = this.destroyWhenButtonClicked ? () => {this.destroy();} : () => {};
+            buttonNode.addEventListener("click", (event) => {
```
Simplify arrow functions
```suggestion
            const destroyFromButton = this.destroyWhenButtonClicked ? () => this.destroy() : () => {};
            buttonNode.addEventListener("click", event => {
```

psvenk commented on public/js/home.js:
```diff
@@ -1181,6 +1192,220 @@ function closeSideNav() {
+            //waits a momement to make sure the snackbar's animation functions properly
+            setTimeout(removeHidden, 10);
+            return this;
+
```
Delete extraneous blank line
```suggestion
```

psvenk commented on public/js/home.js:
```diff
@@ -1181,6 +1192,220 @@ function closeSideNav() {
+        //function for the callback for setTimeout
+        const removeHidden = function() {
+            snackbar.element.classList.remove("hidden");
+        }
```
I think this can be simplified while still paying attention to scope (untested):
```suggestion
        const removeHidden = () => this.element.classList.remove("hidden");
```

psvenk commented on public/js/home.js:
```diff
@@ -1181,6 +1192,220 @@ function closeSideNav() {
+
+        return id;
+    }
+
```
Delete extraneous blank line
```suggestion
```

psvenk commented on public/js/home.js:
```diff
@@ -1181,6 +1192,220 @@ function closeSideNav() {
+
+        //makes the button if given button parameters
+        if (this.buttonText !== undefined && this.buttonClick != undefined) {
+
```
Delete extraneous blank line
```suggestion
```

andOrlando committed:
Apply suggestions from code review

Co-authored-by: psvenk <45520974+psvenk@users.noreply.github.com>

kdk1616 commented:
Add removed assignments to a stack and then they can be returned with undo button (or Cmd+z)

andOrlando committed:
fixes and improves the timer

- fixes bug where the timeout wouldn't run the correct function on timeout
- fixes bug where the timeout would start when the snackbar is made
- adds timeout modes for what to do when the timer finishes

andOrlando committed:
fixes bug with new assignments

- gives new assignments sequential IDs
- filters the assignments when determing whether to show the info icon or not
- changes right padding for button

andOrlando committed:
Update public/js/home.js

I don't know how to get rid of this suggestion otherwise so I'll cave

Co-authored-by: psvenk <45520974+psvenk@users.noreply.github.com>

notrodes reviewed (dismissed):
split off into file

andOrlando committed:
moves snackbar code to another file

andOrlando committed:
adds an undo feature with ctrl+z

-fixes a random tooltips bug for some reason
-changes the ::after to ::before for no reason
-adds an undo-registry to keep track of all the undos
-adds states to make it easier to ascertain the status of a given snackbar
-slightly improves a couple comments

andOrlando committed:
removes console.logs

I left them in for debugging...

andOrlando committed:
Fixes snackbar bug

Rejoice! After a whole month or something I finally fixed that one bug!
(changed method of keeping track of snackbar stuff)

andOrlando committed:
Fixes undo order

I was using unshift instead of shift which wasn't great, the new order
feels much more intuitive

psvenk committed:
Normalize whitespace and fix spelling

psvenk reviewed (changes requested):
Looks good overall. A few comments:
- Deleting an assignment or clicking "undo" in the snackbar scrolls up to the top of the page
- Clicking anywhere on the snackbar other than the "undo" button dismisses the snackbar permanently
- Make Ctrl+Z more discoverable and make it use Command+Z on macOS

And some style nitpicks:
- There are many overlong lines that should be broken up so that they are (for the most part) less than 80 characters long
  + I know that some people question the value of having a rigid limit, but some of these lines do many things at once and are more difficult to read, especially for people who use 80-column windows for editing code (we still exist!)
- Comments should have a space between the comment delimiter and the text (`// Comment`)
- Some statements are missing semicolons

psvenk committed:
Normalize style

Keep lines to <=80 characters, simplify some code, fix indentation
